extensions/transferto/client/client.go

```diff
@@ -0,0 +1,218 @@
+	OriginatingCurrency string          `json:"originating_currency"`
+	ProductRequested    decimal.Decimal `json:"product_requested"`
+	ActualProductSent   decimal.Decimal `json:"actual_product_sent"`
+	SMSSent             string          `json:"sms_sent"`
+	SMS                 string          `json:"sms"`
+	WholesalePrice      decimal.Decimal `json:"wholesale_price"`
+	ServiceFee          decimal.Decimal `json:"service_fee"`
+	RetailPrice         decimal.Decimal `json:"retail_price"`
+	LocalInfoAmount     decimal.Decimal `json:"local_info_amount"`
+	LocalInfoValue      decimal.Decimal `json:"local_info_value"`
+	Balance             decimal.Decimal `json:"balance"`
+}
+
+// Topup makes an actual airtime transfer
+func (c *Client) Topup(reservedID int, msisdn string, destinationMSISDN string, product string, skuid string) (*Topup, error) {
+	request := url.Values{}
+	request.Add("action", "topup")
+	request.Add("reserved_id", strconv.Itoa(reservedID))
+	request.Add("msisdn", msisdn)
+	request.Add("destination_msisdn", destinationMSISDN)
+	request.Add("product", product)
+	if skuid != "" {
+		request.Add("skuid", skuid)
+	}
+
+	response := &Topup{}
+	if err := c.request(request, response); err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+// makes a request with the given data and parses the response into the destination struct
+func (c *Client) request(data url.Values, dest interface{}) error {
+	key := strconv.Itoa(int(time.Now().UnixNano() / int64(time.Millisecond)))
+
+	hasher := md5.New()
+	hasher.Write([]byte(c.login + c.token + key))
+	hash := hex.EncodeToString(hasher.Sum(nil))
+
+	data.Add("login", c.login)
+	data.Add("key", key)
+	data.Add("md5", hash)
+
+	req, err := http.NewRequest("POST", c.apiURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	response, _, err := c.httpClient.DoWithDump(req)
+	if err != nil {
+		return err
+	}
+
+	defer response.Body.Close()
+	if err := c.parseResponse(response.Body, dest); err != nil {
+		return fmt.Errorf("transferto API call returned an unparseable response: %s", err)
+	}
+
+	baseResp := dest.(Response)
+	if baseResp.ErrorCode() != 0 {
+		return fmt.Errorf("transferto API call returned an error: %s (%d)", baseResp.ErrorTxt(), baseResp.ErrorCode())
+	}
+	return nil
+}
+
+// reads and parses a response body, which is in the format
+//
+// value1=result1
+// value2=result2
+// ...
+//
+// with each line separated by \r\n
+func (c *Client) parseResponse(body io.Reader, dest interface{}) error {
+	asBytes, err := ioutil.ReadAll(body)
+	if err != nil {
+		return err
+	}
+
+	// parse into a map
+	data := make(map[string]string)
+	for _, line := range strings.Split(string(asBytes), "\r\n") {
+		parts := strings.SplitN(line, "=", 2)
+
+		if len(parts) == 2 {
+			data[parts[0]] = parts[1]
+		}
+	}
+
+	// marshal to JSON and then into the destination struct
+	respJSON, _ := json.Marshal(data)
```

Comment:
this is a little weird - parse the custom response format, into a map, marshal to JSON, and then leverage golang's JSON unmarshaling to get that into our structs. I tried using the mapstructure package instead but it doesn't support custom type specific unmarshaling. Could probably rewrite all of this to XML and golang's builtin XML marshalling.